torchao/prototype/low_bit_optim/subclass_4bit.py

```diff
@@ -38,10 +39,7 @@ def __init__(self, codes: Tensor, scale: Tensor, qmap: Tensor, signed: bool, sha
         self.qmap = qmap
         self.signed = signed
         self._shape = shape
-
-    @property
-    def block_size(self):
-        return self.codes.numel() * 2 // self.scale.numel()
+        self.block_size = codes.numel() * 2 // scale.numel()
```

Comment:
curious q: Is there some description of the codes/ scales tensor and their relation to each other?

I can see the pattern that codes has .5x (4bit) and 1x (8bit) the bsize * scale numels
But does this assert square blocks?
I think some description here would be helpful

Reply:
I will add some description. Basically for 8-bit and FP8, `codes` has the same shape as the "outer shape", while for 4-bit, since there is bit-packing, I find that it's easier to let `codes` be a flattened 1D buffer and keep track of the shape manually.
To get the scale, the float tensor is actually flattened first and reshape to `(-1, block_size)`. This is done to relax the requirement that the last dimension must be divisible by `block_size` -> now we only need numel (total size) to be divisible by `block_size`. This is especially needed when block size is large (8-bit optim uses block_size=2048 as done in bnb). Since optim update is element-wise, we don't really need to care if the original tensor is 1D, 2D, or n-D (well, maybe there is some structure in n-D tensor that flattening it might not be so wise). I believe the original implementation in bnb does this as well.
-> `scale` is always a 1D tensor, with `size=original_tensor.numel() // block_size`

Reply:
@drisspg Added some docs. Lmk if it is still unclear.

Reply:
Done